File: java/src/main/java/nz/sodium/time/SecondsTimerSystemImpl.java
package nz.sodium.time;

class SecondsTimerSystemImpl implements TimerSystemImpl<Double> {
    public SecondsTimerSystemImpl() {
        this(new MillisecondsTimerSystemImpl());
    }
    private SecondsTimerSystemImpl(TimerSystemImpl<Long> impl) {
        this.t0 = impl.now();
        this.impl = impl;
    }
    private final long t0;
    private final TimerSystemImpl<Long> impl;
    public Timer setTimer(Double t, Runnable callback) {
        return impl.setTimer((long)(t * 1000.0) + t0, callback);
    }
    public void runTimersTo(Double now) {
        impl.runTimersTo((long)(now * 1000.0) + t0);
    }
    public Double now() {
        return (double)(impl.now() - t0) / 1000.0;
    }
}

File: swift/src/Sodium/Sodium/Sodium/Sodium/Time/SecondsTimerSystemImpl.swift
class SecondsTimerImpl: TimerSystemImpl<Double> {
}

class TimerSystemImpl<T>: ITimerSystemImpl {
    
    convenience init() {
        self.init(impl: MillisecondsTimerSystemImpl())
    }
    
    private init(impl: ITimerSystemImpl) {
        self.t0 = impl.now
        self.impl = impl
    }
    private let t0 = T
    private let impl: ITimerSystemImpl
    
    // Mark: ITimerSystemImpl
    func setTimer(t: T, callback: Action) -> Timer {
        return impl.setTimer((T)(t * 1000.0) + t0, callback)
    }
    
    public func runTimersTo(now: T) {
        impl.runTimersTo((T)(now * 1000.0) + t0)
    }
    
    public var now: T {
        return (T)(impl.now - t0) / 1000.0
    }
}

